data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,161 @@
+import random
+import sys
+
+from time import time_ns
+
+from google.cloud import bigquery
+
+from py4j.protocol import Py4JJavaError
+from pyspark.sql import SparkSession
+
+from pyspark.sql.functions import UserDefinedFunction
+from pyspark.sql.types import IntegerType, StringType
+
+
+BUCKET_NAME = sys.argv[1]
+TABLE = "bigquery-public-data.new_york_citibike.citibike_trips"
+
+
+# START MAKING DATA DIRTY
+def random_select(items, weights):
+    '''Picks an item according to the cumulative weights'''
+    return random.choices(items, weights=weights, k=1)[0]
+
+
+def trip_duration(duration):
+    '''Converts trip duration to other units'''
+    seconds = str(duration) + " s"
+    minutes = str(float(duration) / 60) + " min"
+    hours = str(float(duration) / 3600) + " h"
+    return random_select([seconds, minutes, hours,
+                         str(random.randint(-1000, -1))],
+                         [0.3, 0.3, 0.3, 0.1])
+
+
+def station_name(name):
+    '''Replaces '&' with '/' with a 50% chance'''
+    return random.choice([name, name.replace("&", "/")])
+
+
+def user_type(user):
+    '''Manipulates the user type string'''
+    return random.choice([user, user.upper(), user.lower(),
+                          "sub" if user == "Subscriber" else user,
+                          "cust" if user == "Customer" else user])
+
+
+def gender(s):
+    '''Manipulates the gender string'''
+    return random.choice([s, s.upper(), s.lower(),
+                         s[0] if len(s) > 0 else "",
+                         s[0].lower() if len(s) > 0 else ""])
+
+
+def convert_angle(angle):
+    '''Converts long and lat to DMS notation'''
+    degrees = int(angle)
+    minutes = int((angle - degrees) * 60)
+    seconds = int((angle - degrees - minutes/60) * 3600)
+    new_angle = str(degrees) + u"\u00B0" + \
+        str(minutes) + "'" + str(seconds) + '"'
+    return random_select([str(angle), new_angle], [0.55, 0.45])
+
+
+# This function is nested since a UserDefinedFunction is
+# expected to take a single argument
+def dirty_data(proc_func, allow_none):
+    '''Master function returns a user defined function
+    that transforms the column data'''
+    def udf(col_value):
+        random.seed(hash(col_value) + time_ns())
+        if col_value is None:
+            return col_value
+        elif allow_none:
+            return random_select([None, proc_func(col_value)],
+                                 [0.05, 0.95])
+        else:
+            return proc_func(col_value)
+    return udf
+
+
+def id(x):
```

Comment:
nit - if this is bike_id, can we call it bike id? id is very generic otherwise

Reply:
I would also add that there should be a brief comment of why a function that just returns its input value is necessary.